### PR TITLE
[ASA] Add week 3 content

### PR DESCRIPTION
**Pull request by randomicecube**
To-do List:

- [X] Topological Ordering
- [X] Strongly Connected Components
- [X] Start BFS

### CODE REVIEW

**Comment by randomicecube** [comment]
@diogotcorreia forgot to add the slides and teacher's notes - won't be able to add them right now, please don't merge the PR while I don't add them.

**Review by diogotcorreia — CHANGES_REQUESTED** [review]
Not very important changes. I think everything else is great, but I've only skimmed through it.
Feel free to merge it anyway, it's good enough for now and it can be improved later if needed.

File: content/asa/0004-algoritmos-elementares.md [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+
+## DFS - _Depth First Search_
+
+Consiste num algoritmo recursivo que utiliza a noção de _backtracking_. Procura esgotar todos os caminhos possíveis em frente, voltando para trás quando já não há mais possibilidades de seguir em frente. Foi abordado em IAED, e será aprofundado nesta secção.
+
+:::details[Exemplo simples da execução do algoritmo]
+
+Consideremos o seguinte grafo:
+
+![DFS - Grafo do primeiro exemplo](./assets/0004-dfs-grafo-primeiro-exemplo.png#dark=1)
```
[the remark this answers is not in the document]
Comment (randomicecube) [reply content/asa/0004-algoritmos-elementares.md:29]: Alright, thanks!

File: content/asa/0004-algoritmos-elementares.md [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+- Do vértice C não temos para onde ir, pelo que o trancamos. (_timestamp_ 12 a "fechar" C)
+
+Aqui, terminamos a procura - não restam vértices por explorar.
+
+A tabela abaixo indica os tempos de descoberta e fecho de cada vértice:
+
+|       | A   | B   | C   | D   | E   | F   |
+| ----- | --- | --- | --- | --- | --- | --- |
+| $d_i$ | 1   | 2   | 3   | 4   | 9   | 10  |
+| $f_i$ | 8   | 7   | 6   | 5   | 12  | 11  |
```
Comment (diogotcorreia) [inline content/asa/0004-algoritmos-elementares.md:64]: While doing the GIF, I noticed this is wrong. C is swapped with E.

**Review by diogotcorreia — CHANGES_REQUESTED** [review]

**Review by diogotcorreia — APPROVED** [review]

**Pull request merged**